File: cpp/84_largest_rectangle_in_histogram.cpp
#include <vector>
#include <stack>

using namespace std;

class Solution {
 public:
  int largestRectangleArea(vector<int>& heights) {
    heights.push_back(0);
    int area = 0;
    stack<int> st;
    for (int right = 0; right < heights.size(); right++) {
      while (!st.empty() && heights[st.top()] > heights[right]) {
        int height = heights[st.top()];
        st.pop();
        int left = (st.empty()) ? -1 : st.top();
        area = max(area, (right - left - 1) * height);
      }
      st.push(right);
    }

    return area;
  }
};

File: rust/84_largest_rectangle_in_histogram.rs
use std::cmp;

impl Solution {
    pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
        let mut area = 0;
        let mut heights = heights;
        heights.push(0);
        let mut st = vec![];

        for right in 0..heights.len() {
            while let Some(&idx) = st.last() {
                if heights[idx] > heights[right] {
                    let height = heights[idx];
                    st.pop();
                    let left = if st.is_empty() {
                        -1
                    } else {
                        *st.last().unwrap() as i32
                    };
                    area = cmp::max(area, (right as i32 - left - 1) * height);
                } else {
                    break;
                }
            }
            st.push(right);
        }

        area
    }
}
